audinterface/core/feature.py

```diff
@@ -301,25 +301,86 @@ def process_index(
+            cache_root: str = None,
     ) -> pd.DataFrame:
         r"""Extract features from an index conform to audformat_.
 
+        If cache folder is not ``None``,
+        a hash value is created from the index
+        using :func:`audformat.utils.hash` and
+        the result is stored as
+        ``<cache_root>/<hash>.pkl``.
+        When called again with the same index,
+        features will be read from the cached file.
+        As a sanity check,
+        columns and values of the first segment are compared.
+        Note that it can still happen
+        that two different processing functions
+        produce the same values for the first segment.
+        We recommend to use unique cache folders
+        for every processing function.
+
         Args:
             index: index with segment information
             root: root folder to expand relative file paths
+            cache_root: cache folder (see description)
 
         Raises:
             RuntimeError: if sampling rates do not match
             RuntimeError: if channel selection is invalid
             RuntimeError: if multiple frames are returned,
                 but ``win_dur`` is not set
+            RuntimeError: if a cached file is found that was
+                generated with a different processing function
             ValueError: if index is not conform to audformat_
 
         .. _audformat: https://audeering.github.io/audformat/data-format.html
 
         """
-        series = self.process.process_index(index, root=root)
-        return self._series_to_frame(series)
+        cache_path = None
+
+        if cache_root is not None:
+            cache_root = audeer.mkdir(cache_root)
+            hash = audformat.utils.hash(index)
+            cache_path = os.path.join(cache_root, f'{hash}.pkl')
+
+        if cache_path and os.path.exists(cache_path):
+
+            df = pd.read_pickle(cache_path)
+
+            if list(df.columns) != self.column_names:
+                raise RuntimeError(
+                    'Found cached file with same index '
+                    'but different columns. '
+                    'Disable caching '
+                    'or choose another cache folder.'
+                )
+
+            if len(index) > 0:
+                df_verify = self.process_index(
+                    index[:1],
+                    root=root,
+                )
+                if not df[:1].equals(df_verify):
+                    raise RuntimeError(
+                        'Found cached file with same index '
+                        'but different values. '
+                        'Disable caching '
+                        'or choose another cache folder.'
+                    )
+
+        else:
+
+            y = self.process.process_index(
+                index,
+                root=root,
+            )
+            df = self._series_to_frame(y)
+
+            if cache_path is not None:
+                df.to_pickle(cache_path, protocol=4)
```

Comment:
As we call `Process.process_index()` inside `Feature.process_index()` anyway, we could also add the caching directly to `audinterface.Process` and simply hand the `cache_root` when calling `self.process.process_index()`.

The advantage would be that we would then support `Process` and `Feature` with the same cache folder.
The downside is that we would need to call `self._series_to_frame(y)` for `Feature` every time we load from cache.
But in my opinion the advantage of supporting both would be bigger.

Reply:
Actually, that's how I implemented it in the first place, but as you said there is the downside that we need to call `self._series_to_frame(y)`. The other disadvantage is that a user then cannot directly load the `DataFrame` from the cache. And personally, I use `Feature` in 99% of the times and almost never `Process`. So if we want to have the cache feature with `Process` I'd suggest we add it separately.

Reply:
>  I use Feature in 99% of the times and almost never Process

For me it's more the opposite as `Process` allows you:

```python
df['prediction'] = model.process_index(df.index)
```

But it doesn't matter, I would simply say we should support the cache in both.

> So if we want to have the cache feature with Process I'd suggest we add it separately.

Yes, but they will try to load the same PKL file as the file name is only given by the index. Will this also work if we go with your solution here and then later on add the same to `Process`?

Reply:
> The other disadvantage is that a user then cannot directly load the DataFrame from the cache

It's still a valid PKL file and the user can load it. It will just contain the raw predictions of the model, but I don't see anything wrong with this.

Reply:
> It's still a valid PKL file and the user can load it. It will just contain the raw predictions of the model, but I don't see anything wrong with this.

1. returns a `pd.Series` instead of a `pd.DataFrame`
2. does not contain feature names
3. in case of framewise output there is no correct time information for the frames

Reply:
> Yes, but they will try to load the same PKL file as the file name is only given by the index.

You have to use a different cache folder. But you should anyway use a different one for every interface.

Reply:
I removed all sanity checks. And as you said, the caching functionality we add here is not meant to be used as a shared cache (therefore we have `audfeatures`). And also the user is responsible to take care of not loading a cache file generated with a different processing function.

Reply:
I just removed the check :)

And I would certainly not replace the feature names silently, as it's a strong sign that the values are also not the desired once.

Frankly, I would simply stay with the solution now that we have no checks.

Reply:
If you don't like renaming the feature column, you could still raise an error as before, but then the error message should list how the expected feature names are.

Reply:
The main danger I see is that you have model A and model B, which both output the same classes and that you accidentally use the same cache folder. In that case, comparing the number of columns and feature names will not help. That's why I had initially added the second check that the values of the first segment have to match. So I would either add both checks or no checks at all.

Reply:
If we stay with the current solution, then we might get the following problem.

Let's assume we start with the following code

```python
feature = 'gender'
interface = audinterface.Feature([feature])
df = interface.process_index(index, cache_root='~/cache/model_id')
df[feature]
```

Now at some point we decide the feature should be called sex instead of gender and modify the code to

```python
feature = 'sex'
interface = audinterface.Feature([feature])
df = interface.process_index(index, cache_root='~/cache/model_id')
df[feature]
```

This will simply fail and from just reading the code it will be hard to find the error.

Reply:
If we decide to also support caching to `Process` we also cannot check for columns.